kvserver,kvnemesis: dump raft logs on failure

tbg opened the pull request:
This helps diagnose failures such as #146043 and #143317.

```
$ go test [...]
somewhere.go:6207: dumping raft logs to [log-scope]/raftlog
```

```
$ ls [log-scope]/raftlog
r1.txt	r12.txt	r15.txt [...]
```

```
$ tail -n 10 r74.txt
****** index 25 ******
RaftCommand: proposer_lease_sequence:1 max_lease_index:20 closed_timestamp:<wall_time:1746450987106072000 > replicated_eval_result:<write_timestamp:<wall_time:1746450990105858000 > delta:<sys_bytes:147 > > logical_op_log:<>
Put: /Local/Lock/Local/Range/Table/72/RangeDescriptor 032e9bd313c26c4cb99422657288fdc2a8 (0x017a6b12016b12d000ff0172647363000100032e9bd313c26c4cb99422657288fdc2a812): 1746450990.105858000,0 txn={id=2e9bd313 key=/Local/Range/Table/72/RangeDescriptor iso=Serializable pri=0.00381679 epo=0 ts=1746450990.105858000,0 min=1746450990.105858000,0 seq=1} ts=1746450990.105858000,0 del=false klen=12 vlen=49 mergeTs=<nil> txnDidNotUpdateMeta=true
Put: 1746450990.105858000,0 /Local/Range/Table/72/RangeDescriptor (0x016b12d000017264736300183ca3ec22e38fd009): [/Table/72, /Table/Max)
	Raw:r74:/Table/{72-Max} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=5]
****** index 26 ******
RaftCommand: proposer_lease_sequence:1 max_lease_index:21 closed_timestamp:<wall_time:1746450987106175000 > replicated_eval_result:<write_timestamp:<wall_time:1746450990105858000 > delta:<sys_bytes:175 sys_count:1 > > logical_op_log:<>
Put: /Local/Lock/Local/Range/Table/Max/RangeDescriptor 032e9bd313c26c4cb99422657288fdc2a8 (0x017a6b12016b12fa00ff0172647363000100032e9bd313c26c4cb99422657288fdc2a812): 1746450990.105858000,0 txn={id=2e9bd313 key=/Local/Range/Table/72/RangeDescriptor iso=Serializable pri=0.00381679 epo=0 ts=1746450990.105858000,0 min=1746450990.105858000,0 seq=2} ts=1746450990.105858000,0 del=false klen=12 vlen=66 mergeTs=<nil> txnDidNotUpdateMeta=true
Put: 1746450990.105858000,0 /Local/Range/Table/Max/RangeDescriptor (0x016b12fa00017264736300183ca3ec22e38fd009): [/Table/Max, /Max)
	Raw:r75:/{Table/Max-Max} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=5, sticky=9223372036.854775807,2147483647]
```

Closes https://github.com/cockroachdb/cockroach/issues/146043

Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146084)
<!-- Reviewable:end -->

tbg committed:
kvserverpb: add getter for WriteBatch

tbg committed:
print: add DumpRaftLog

A note here: import cycles are quite bad with the print package. Since it
depends on "everything" (storage, raftlog, etc) it's not usable from any of
these packages. Annoyingly, the more powerful the print package becomes, the
less likely it is that one can simply use it in tests that want it (unless
they're in `_test`, of course).

Ideally we'd convert the print package to a "registration" pattern, where most
packages inject the relevant dependencies into `print`. However, `print` heavily
depends on types from the `storage` package, so it won't be usable from
`storage` even with that, and a bigger refactor would be necessary.

tbg committed:
kvserver: dump raft log on failure in TestMergeReplicatesLocks

Closes #146043.

tbg committed:
kvnemesis: dump raft logs on failure

See #143317.

tbg committed:
kvtestutils: add RaftLogDumper

It will be used in the next commit.

tbg committed:
kvserver: skip TestMergeReplicatesLocks under duress

stevendanna commented on pkg/kv/kvserver/print/dump_raft_log.go:
```diff
@@ -0,0 +1,57 @@
+	return raftlog.Visit(ctx, reader, rangeID, 0, math.MaxUint64, func(ent raftpb.Entry) error {
+		put("****** index %d ******", ent.Index)
+		e, err := raftlog.NewEntry(ent)
+		defer e.Release()
```
Is this valid to release in the case of error?

tbg replied:
No, good catch, fixed with next push.

stevendanna commented on pkg/kv/kvserver/print/dump_raft_log.go:
```diff
@@ -0,0 +1,57 @@
+		s, err := DecodeWriteBatch(wb.GetData())
+		if err != nil {
+			put("%s", err)
+			// continue
```
Did you intend to return here?

tbg replied:
Yep, done

stevendanna commented on pkg/kv/kvserver/client_replica_test.go:
```diff
@@ -6250,7 +6250,10 @@ func TestMergeDropsLocksIfLargerThanMax(t *testing.T) {
+	scope := log.Scope(t)
+	defer scope.Close(t)
+
+	skip.UnderDuress(t, "too slow for testrace")
```
I can try to follow up here. We can probably get away with a single-node test for this which will probably help a bit.

stevendanna reviewed (commented):
Thanks for working on this! Just a few minor comments.

tbg committed:
print: fix buglets

tbg commented:
TFTR!

bors r+
